x-pack/legacy/plugins/siem/server/lib/detection_engine/signals/build_exceptions_query.ts

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { isEmpty } from 'lodash';
+
+import { Query } from '../../../../../../../../src/plugins/data/server';
+import { List, ListOperator, ListValues } from '../routes/schemas/types/lists_default_array';
+import { RuleAlertParams } from '../types';
+
+export const getLanguageBooleanOperator = (language: string, value: string) => {
+  if (language === 'lucene') {
+    return value.toUpperCase();
+  }
+
+  return value;
+};
+
+export const operatorBuilder = (operator: ListOperator, language: string): string => {
+  return operator === 'excluded'
+    ? ` ${getLanguageBooleanOperator(language, 'and')} `
+    : ` ${getLanguageBooleanOperator(language, 'and')} ${getLanguageBooleanOperator(
+        language,
+        'not'
+      )} `;
+};
+
+export const buildExists = ({
+  operator,
+  field,
+  language,
+}: {
+  operator: ListOperator;
+  field: string;
+  language: string;
+}): string => {
+  switch (language) {
+    case 'kuery':
+      return `${operatorBuilder(operator, language)}${field}:*`;
+    case 'lucene':
+      return `${operatorBuilder(operator, language)}_exists_${field}`;
+    default:
+      return '';
+  }
+};
+
+export const buildMatch = ({
+  operator,
+  field,
+  values,
+  language,
+}: {
+  operator: ListOperator;
+  field: string;
+  values: ListValues[];
+  language: string;
+}) => {
+  const value = values[0].name;
+  return `${operatorBuilder(operator, language)}${field}:${value}`;
+};
+
+export const buildMatchAll = ({
+  operator,
+  field,
+  values,
+  language,
+}: {
+  operator: ListOperator;
+  field: string;
+  values: ListValues[];
+  language: string;
+}) => {
+  if (values.length === 1) {
+    return buildMatch({ operator, field, values, language });
+  } else {
+    const matchAllValues = values.map(value => {
+      return value.name;
+    });
+
+    return `${operatorBuilder(operator, language)}${field}:(${matchAllValues.join(
+      ` ${getLanguageBooleanOperator(language, 'or')} `
+    )})`;
+  }
+};
+
+export const evaluateValues = ({ list, language }: { list: List; language: string }): string => {
+  const { values_operator: operator, values_type: type, field, values } = list;
+  switch (type) {
+    case 'exists':
+      return buildExists({ operator, field, language });
+    case 'match':
+      return buildMatch({ operator, field, values: values!, language });
+    case 'match_all':
+      return buildMatchAll({ operator, field, values: values!, language });
+    default:
+      return '';
+  }
+};
+
+export const buildExceptions = ({
+  query,
+  lists,
+  language,
+  accumulator = [],
```

Comment:
I don't see this fourth argument ever being used? Can we remove it? It's also dangerous to mutate the input parameters such as an array as both the input and output. The way this is written it looks like maybe if someone passed in an accumulator it would equal the output?

I tested this and it looks like it works:

```ts
export const buildExceptions = ({
  query,
  lists,
  language,
}: {
  query: string;
  lists: List[];
  language: string;
}): string[] => {
  return lists.reduce<string[]>((acc, item) => {
```

Notice how I do:

```ts
  return lists.reduce<string[]>((acc, item) => {
```

That makes it so I that TypeScript is happy since I tell it I am going to accumulate over a string array.

Reply:
Thanks for pointing that out! That was leftover from a different iteration of that function where I was using recursion and passing through a different accumulator. That isn't being used anywhere in the code. I updated with your suggested code.